### ISSUE TITLE
When train the model with Lora, the ref_model in npo is None.

### ISSUE THREAD

**Issue by sev777** [jinzhuoran/RWKU#7]
https://github.com/jinzhuoran/RWKU/blob/b8a03b3ce34fb4a96001df545a56558d75a078a3/LLaMA-Factory/src/llmtuner/train/npo/trainer.py#L61

`
    forget_outputs_oracle = self.ref_model(inputs['input_ids'], labels=inputs['labels'], attention_mask=inputs['attention_mask'])
`
How to init the ref_model in NPO with lora?

`
def create_ref_model(
    model_args: "ModelArguments", finetuning_args: "FinetuningArguments", add_valuehead: bool = False
) -> Optional[Union["PreTrainedModel", "AutoModelForCausalLMWithValueHead"]]:
    r"""
    Creates reference model for PPO/DPO training. Evaluation mode is not supported.

    The valuehead parameter is randomly initialized since it is useless for PPO training.
    """
    if finetuning_args.ref_model is not None:
        ref_model_args_dict = model_args.to_dict()
        ref_model_args_dict.update(
            dict(
                model_name_or_path=finetuning_args.ref_model,
                adapter_name_or_path=finetuning_args.ref_model_adapters,
                quantization_bit=finetuning_args.ref_model_quantization_bit,
            )
        )
        ref_model_args = ModelArguments(**ref_model_args_dict)
        ref_finetuning_args = FinetuningArguments(finetuning_type="lora")
        tokenizer = load_tokenizer(ref_model_args)["tokenizer"]
        ref_model = load_model(
            tokenizer, ref_model_args, ref_finetuning_args, is_trainable=False, add_valuehead=add_valuehead
        )
        logger.info("Created reference model from {}".format(finetuning_args.ref_model))
    else:
        if finetuning_args.finetuning_type == "lora":
            ref_model = None
        else:
            tokenizer = load_tokenizer(model_args)["tokenizer"]
            ref_model = load_model(
                tokenizer, model_args, finetuning_args, is_trainable=False, add_valuehead=add_valuehead
            )
            logger.info("Created reference model from the model itself.")

    return ref_model
`

**Comment by Acruxos** [comment]
Hi, I'm not sure if you've already solved this issue. I referred to the implementation of DPO and modified the compute_loss function in `npo/trainer.py` like this, which seems to have solved the problem. 

Posting this here to save time for anyone encountering the same issue:

```python
    def compute_loss(self, model, inputs, return_outputs=False):
        loss, outputs = super().compute_loss(model, inputs, True)
        forget_loss_current = self.get_batch_loss(outputs.logits, inputs['labels'])
        with torch.no_grad():
            if self.ref_model is None:
                ref_model = self.model
                ref_context = self.accelerator.unwrap_model(self.model).disable_adapter()
            else:
                ref_model = self.ref_model
                ref_context = nullcontext()

            with ref_context:
                forget_outputs_oracle = ref_model(inputs['input_ids'], labels=inputs['labels'], attention_mask=inputs['attention_mask'])
            forget_logits_oracle = forget_outputs_oracle.logits
            forget_loss_oracle = self.get_batch_loss(forget_logits_oracle, inputs['labels'])
        neg_log_ratios = forget_loss_current - forget_loss_oracle
        loss = -F.logsigmoid(self.beta * neg_log_ratios).mean() * 2 / self.beta
        return (loss, outputs) if return_outputs else loss
```

When ref_model is None, the original model is used as ref_model within a context that disables the LoRA modules. It's strange that this was missing in the original code.

**Comment by scvready123** [comment]
Hi, I met the same problem. Have you fixed it? Thx

**Comment by jinzhuoran** [comment]
> Hi, I met the same problem. Have you fixed it? Thx

Sorry for the late reply. Since LoRA does not update the parameters of the original model, we can directly use the original model as the reference model.
